source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,297 @@
+
+These tools usually provides:
+
+* Library to improve the event handling logic: "smart" events queue that dedup events, filtering events, ...
+* A code generator component that, starting from the CRD defined in the code, scaffolds the controller code and the client code to interact with the CRD (this is because Golang doesn't have generics)
+
+Apart the differences between the tools, they all share the same problem: one CRD maps to one controller that maps to one process that maps to one pod. Every time we want to extend Kubernetes with a new API we add a new process which generates network traffic and waste CPU and memory, even when it's just listening. 
+In short, **We're flooding our Kubernetes clusters with applications that suck resources and remain in idle for most of their time!**
+
+Knative community, with some well crafted engineering, found a solution to package different controllers **at compile time** in the same process. This is fine, but that's still not a solution to the initial problem: we cannot build Kubernetes every time packaging all the controllers we need, we want a solution to dynamically extend the Kubernetes APIs.
+
+## Extending Kubernetes APIs in process
+
+The idea behind our solution is to create a runtime plugin system where every plugin is a Kubernetes controller.
+
+{%fi /images/kubernetes-controllers-a-new-hope/kubernetes-controllers-a-new-hope-1.png, Plugin system, Plugin system %}
+
+This plugin system should have certain traits:
+
+* Plugins should be well isolated between them aka no plugin failure/bad behaviour should annoy other plugins
+* The host should be able to control and restrict what the plugin can do
+* Users should be able to develop plugins with different languages
+
+This plugin system may live in a _mega controller_, or it may live directly inside the _kube-controller-manager_, which is the process with all the controllers that manage the Kubernetes built-in building blocks:
+
+{%fi /images/kubernetes-controllers-a-new-hope/kubernetes-controllers-a-new-hope-2.png, kube-controller-manager including the plugin system, kube-controller-manager including the plugin system %}
+
+## Isolated, multi-language, lightweight plugin system? That sounds hard...
+
+The requirements of the plugin system are not trivial, but lucky for us WebAssembly comes in rescue!
+
+Wasm is an instruction set, targetable from every programming language, that runs in an isolated Virtual Machine. From [webassembly.org](https://webassembly.org/): _"WebAssembly (abbreviated Wasm) is a binary instruction format for a stack-based virtual machine. Wasm is designed as a portable compilation target for programming languages, enabling deployment on the web for client and server applications"_. That sounds like the perfect fit for our idea:
+
+* Because is isolated by default, we define piece by piece all the possible interactions between the plugins (controllers) and the host
```

Comment:
```suggestion
* Because it's isolated by default, we define all the possible interactions between the plugins (controllers) and the host, piece by piece
```